File: frontend/src/types/alerts.ts
export interface AlertRecord {
  id?: number;
  alert_id: string;
  timestamp: string;
  severity: string;
  title: string;
  description?: string;
  source_event_id?: string | null;
  entity_id?: string | null;
  event_type?: string | null;
  source_topic?: string;
  anomaly_score?: number | null;
  ml_risk_score?: number | null;
  correlation_id?: string;
  incident_id?: string | null;
  explanation: string;
  recommended_action: string;
  status: string;
  tags: string[];
}

export interface IncidentRecord {
  id?: number;
  incident_id: string;
  created_at: string;
  updated_at: string;
  severity: string;
  title: string;
  description: string;
  related_alerts: string[];
  entity_ids: string[];
  status: string;
  event_count: number;
  escalation_level: number;
  resolution_notes?: string | null;
  raw_payload?: Record<string, unknown>;
}

export interface AlertStats {
  total_alerts: number;
  open_alerts: number;
  critical_alerts: number;
  high_alerts: number;
  total_incidents: number;
  open_incidents: number;
  severity_counts: Record<string, number>;
}

export interface AlertsLiveMessage {
  timestamp: string;
  recent_alerts: AlertRecord[];
  critical_alerts: AlertRecord[];
  recent_incidents: IncidentRecord[];
  severity_counters: Record<string, number>;
  stats: AlertStats;
}

File: backend/app/schemas/alerts.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict

AlertStatus = Literal["open", "acknowledged", "investigating", "resolved"]


class AlertResponse(BaseModel):
    id: int
    alert_id: str
    timestamp: datetime
    severity: str
    title: str
    description: str
    source_event_id: str | None = None
    entity_id: str | None = None
    event_type: str | None = None
    source_topic: str
    anomaly_score: float | None = None
    ml_risk_score: float | None = None
    correlation_id: str
    incident_id: str | None = None
    explanation: str
    recommended_action: str
    status: str
    tags: list[str]
    raw_payload: dict[str, Any]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class IncidentResponse(BaseModel):
    id: int
    incident_id: str
    created_at: datetime
    updated_at: datetime
    severity: str
    title: str
    description: str
    related_alerts: list[str]
    entity_ids: list[str]
    status: str
    event_count: int
    escalation_level: int
    resolution_notes: str | None = None
    raw_payload: dict[str, Any]

    model_config = ConfigDict(from_attributes=True)


class StatusUpdateRequest(BaseModel):
    status: AlertStatus
    resolution_notes: str | None = None


class AlertStatsResponse(BaseModel):
    total_alerts: int
    open_alerts: int
    critical_alerts: int
    high_alerts: int
    total_incidents: int
    open_incidents: int
    severity_counts: dict[str, int]


class AlertsLiveMessage(BaseModel):
    timestamp: datetime
    recent_alerts: list[dict[str, Any]]
    critical_alerts: list[dict[str, Any]]
    recent_incidents: list[dict[str, Any]]
    severity_counters: dict[str, int]
    stats: AlertStatsResponse

